### ISSUE TITLE
Location state

### ISSUE THREAD

**Issue by Namielusi** [typehero/type-route#29]
Hello, thanks for this great library, I love how it makes everything simple.

But I wonder if there the not documented possibility to manage location state? I tried to use history API manually, but it turned out that I'm losing connection with the library and things just get complicated.

I think it would be just great if I could pass the state while navigating programmatically (using Type Route API, not history API), and have access to the state as well inside the listener.

Is this possible now, or are there any plans to make it possible? Maybe I can help with that?

**Comment by bradenhs** [comment]
Is this the capability what you are referring to? https://www.type-route.org/docs/guides/programmatic-navigation. Perhaps the docs could use an example on that page to make things more clear. There are examples on the `push` and `replace` docs pages:  https://www.type-route.org/docs/api-reference/route-definition/push, https://www.type-route.org/docs/api-reference/route-definition/replace.

**Comment by Namielusi** [comment]
Push & replace methods provided by the library are similar to push & replace from the history API, but in the first case, I can only pass params (path.param or query.param), which will be used in the URL. But I can't provide the state of the new route.

Just to make it clear, I'm sure you've seen it, but here's the reference to the push & replace from the history API: https://github.com/ReactTraining/history/blob/master/docs/Navigation.md

Particularly, it mentions the possibility to pass the state (as the second parameter or inside location object). The type route library doesn't have the possibility to pass this state object at all, but it uses location state internally, as far as I understood from the source code.

Currently, I've come up with an ugly solution. I use history API directly (push & replace functions) and just write the current route (which I get by using getCurrentRoute function) in the state. Then, I'm using the listener from the history API, getting the route from state and calling setRoute function. This way, I can pass any data I want in the location state (along with the route data).

---

Here's my solution:
```js
import { getCurrentRoute, history, routes } from "./router";

// I create the wrapper for the push/replace functions
const pushEx = (route, params) => {
  const historyInstance = history.getActiveInstance();
  const state = {
    route: { name: route.name, params, action: "push" },
    also: 'any other payload I want to write in the location state'
  };

  historyInstance.push(route.href(params), state);
};

// Then I use it somewhere programmatically
const handleClick = (id) => {
  pushEx(routes.todoDetails, { id })
}

const App = () => {
  const historyInstance = history.getActiveInstance();
  const [route, setRoute] = useState(getCurrentRoute());

  useEffect(
    // I use listener from the History API
    historyInstance.listen(nextLocation => {
      // Where I have access to the state
      const payload = nextLocation.state.also
      setRoute(nextLocation.state.route);
    }),
    []
  );

  return <Page route={route} />
};
```


And here's how it could be:
```js
import { getCurrentRoute, listen, routes } from "./router";

// Use library API without any wrappers
// In this case, you don't need to use the history API at all
const handleClick = (id) => {
  routes.todoDetails.push({ id }, {
    also: 'any other payload I want to write in the location state'
  })
}

const App = () => {
  const [route, setRoute] = useState(getCurrentRoute());

  useEffect(
    // Use listener from the type route library
    listen(nextRoute => {
      // But still have access to the state
      const payload = nextRoute.state.also
      setRoute(nextRoute);
    }),
    []
  );

  return <Page route={route} />
};
```

**Comment by Namielusi** [comment]
I guess the state should have predefined types, like other path/query parameters do. And this way, I think, the state could be passed along with other parameters, just in one object.

```js
routes.todoDetails.push({
  id,
  also: 'any other payload I want to write in the location state'
})
```

My use case requires to have an object placed in the state. So it's not just string/number type. What do you think about it?

**Comment by bradenhs** [comment]
Thanks @Namielusi for going into more detail! I misunderstood the original question.

When building the library including the ability to use the location state part of the history API was considered and it wouldn't be that hard to expand the Type Route API to support it. I haven't personally faced a use case that would merit its inclusion though. Would you mind sharing an example where it would be preferable to use location state as opposed to path and query parameters?

**Comment by Namielusi** [comment]
I think my use-case is not widely spread among projects. I personally store objects that are just too big to store them in URL.

In my case, I need to know the history of pages visited by the user. It's very convenient to use location state for this as it stays even after page reloading, and I don't need to manage my own state as I just write the last 3 pages in the location object. So even if the user uses the back/forward buttons, I still get the history for the particular page.

I can't come up with a use-case other than storing internal-use-only information. Not sure if we should wait for somebody else facing this case. Just to see the need for changes.

**Comment by bradenhs** [comment]
This PR adds an api for state parameters: https://github.com/type-route/type-route/pull/35. Any feedback on it would be appreciated.

**Comment by bradenhs** [comment]
Released support for this in `0.4.0`! Feel free to reopen if you feel that release doesn't address this fully.
